LFS-222 Search for a vocabulary by its acronym or word in its name

prabnur opened the pull request:
Loading bar activates. Search now first looks for vocabularies that match search keywords by name or acronym. Then a request is made to Recommender and the previous search continues. Any results are added without duplicates. After this the loading bar is deactivated.

prabnur committed:
LFS-222 Search for a vocabulary by its acronym or word in its name

prabnur committed:
LFS-222 Remove console.log statement and add detail to a comment

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -75,14 +75,34 @@ export default function Search(props) {
 
-    fetch(vocabLinks["recommender"]["base"]+vocabLinks["apikey"]+"&input="+keywords.replace(" ", "%20"))
+    // Then also make a request to recommender and update filtered list.
+    fetch(vocabLinks["recommender"]["base"] + vocabLinks["apikey"] + "&input="+keywords.replace(" ", "%20"))
```
Why not use the `URL()` and `URL.searchParams` api? It would probably be cleaner.
In addition, the `keywords.replace` should probably instead be a `encodeURIComponent()` call instead

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -75,14 +75,34 @@ export default function Search(props) {
       } else {
         return response;
-      }
+      } 
```
Accidental space?

prabnur replied:
Yea lol

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -75,14 +75,34 @@ export default function Search(props) {
+    let acronymList = [];
+    let keywordsList = keywords.split(" ");
+    keywordsList.forEach(keyword => keyword.toLowerCase());
+    props.vocabList.map((vocab) => {
```
A `map` where you don't use the return value leads to an empty array being instantiated and then destroyed, which is detrimental to performance. Two options here:
- Use a `forEach` instead
- Replace the if statement with a `filter()` call, and then return `vocab.ontology.acronym`, which lets you initialize `acronymList` in one step.

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -75,14 +75,34 @@ export default function Search(props) {
+    fetch(vocabLinks["recommender"]["base"] + vocabLinks["apikey"] + "&input="+keywords.replace(" ", "%20"))
     .then((response) => {
       if (response.status >= 400) {
         errHappened = true;
```
If an error occurs, it's more idiomatic to return `Promise.reject(response)`, which will skip all subsequent `then()` calls. At the end, you should have some sort of error reporting mechanism, like a `<Typography color="error">` element.

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -70,8 +70,20 @@ export default function VocabularyTable(props) {
-          setParentAcronymList={setAcronymList}
-          setParentFilterTable={setFilterTable}
+        concatParentAcronymList={list => {
+          // setAcronymList can also take in a transformation function
```
Might be more readable if instead you write about what it's doing -- namely, that it's giving `Search` a function that allows it to concatenate a new list onto `acronymList`

OrdiNeu commented:
It's looking good, though I'm not sure how to use it properly -- I tried looking for HPO and got the following:

![image](https://user-images.githubusercontent.com/4656440/71041234-96119600-20f6-11ea-9f03-2f22ae993c38.png)

prabnur committed:
LFS-222 Remove redundant maps. Use URL api and encodeURIComponent. Refactor fetch error handling to use returning of Promise.reject(response).

prabnur committed:
LFS-222 Revert install function link generation code. Remove extra newline.

OrdiNeu commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -66,36 +66,59 @@ function extractList(data) {
 
   const classes = useStyles();
 
   function search() {
```
There's a race condition if the user searches two things in quick succession, but it's probably minor.

OrdiNeu reviewed (approved):
Looks good!

OrdiNeu committed:
LFS-222: Search for a vocabulary by its acronym or word in its name
Prevent race condition
Make the "Results containing" subheader only contain what was actually
searched
Add more comments

sdumitriu committed:
LFS-222: Search for a vocabulary by its acronym or word in its name
Bugfix: searching for something that doesn't match should clear the results of the previous search

sdumitriu reviewed (approved):
One thing that I still don't like is that it displays ontologies matching any one word of the query, so searching for "genes and genomes" will display all 25+ ontologies with "and" in their name, while "human phenotype ontology" will show almost all ontologies, with HPO lost in the middle.
